src/coreclr/src/gc/gc.cpp

```diff
@@ -2053,28 +2053,35 @@ ptrdiff_t round_down (ptrdiff_t add, int pitch)
 #error if GROWABLE_SEG_MAPPING_TABLE is defined, SEG_MAPPING_TABLE must be defined
 #endif
 
+// Returns true if moving an object from old_loc to new_loc respects large (double) alignment.
+// old_loc and new_loc are expected to be aligned by DATA_ALIGNMENT.
 inline
-BOOL same_large_alignment_p (uint8_t* p1, uint8_t* p2)
+BOOL same_large_alignment_p (uint8_t* old_loc, uint8_t* new_loc)
 {
 #ifdef RESPECT_LARGE_ALIGNMENT
-    return ((((size_t)p1 ^ (size_t)p2) & 7) == 0);
-#else
-    UNREFERENCED_PARAMETER(p1);
-    UNREFERENCED_PARAMETER(p2);
+    return ((((size_t)old_loc ^ (size_t)new_loc) & DATA_ALIGNMENT) == 0) || (old_loc == 0);
```

Comment:
I would not change & 7 to & DATA_ALIGNMENT because this gets a different answer on 64-bit. I don't think we are defining RESPECT_LARGE_ALIGNMENT on 64-bit but I would still avoid this kind of functional changes.

if you really want, you could add in the comments that by large alignment we really mean 8-byte alignment.

personally I would not do this change at all - now you have to make sure that you pass old_loc as the 1st parameter instead of passing them in either order and just getting an answer about their alignments. but if it really makes you happy I'm ok with it.

also I would do the (old_loc == 0) first -

`return ((old_loc == 0) || ((((size_t)old_loc ^ (size_t)new_loc) & 7) == 0));`

Reply:
Yes, it feels like one of the rare cases where literally `7` seems to fit better, because under RESPECT_LARGE_ALIGNMENT it cannot be anything else. 

Maybe worth adding a compile time assert somwhere that RESPECT_LARGE_ALIGNMENT is used only in 32bit. (or is it only in ARM32 ?)

Reply:
yes I agree about the compile time assert (for 32-bit, doesn't need to be for only ARM32)

Reply:
Why don't we replace `return DATA_ALIGNMENT` with `return 4` in `switch_alignment_size` then?  It does not make sense to use symbolic constant in one function and numeric constant in another.  Large alignment means double `DATA_ALIGNMENT`.  It does not matter if `DATA_ALIGNMENT` is 4 or 8, it will still work.

Reply:
> I would not change & 7 to & DATA_ALIGNMENT because this gets a different answer on 64-bit.

And I think that is good.  Large alignment means double alignment, not 64-bit alignment.  If it was intended to mean 64-bit alignment, then `switch_alignment_size` would do bitwise-or with 4, not with `DATA_ALIGNMENT`.

Reply:
> now you have to make sure that you pass old_loc as the 1st parameter instead of passing them in either order

This is a valid argument; the function name does not reflect this argument asymmetry.  I put `old_loc == 0` at the end for performance, because I expect it to be a rare condition.  On the other hand, I expect `old_loc` and `new_loc` to have the same large alignment in 50% cases.

Reply:
> Why don't we replace return DATA_ALIGNMENT with return 4 in switch_alignment_size then

I'd much rather you make that change than this one if it really makes you feel better 'cause I don't advise to make carelessly functional changes. large alignment means 4-byte vs 8-byte, it's not tested on any other alignments. while your change gives the right answer for what you think it means, it's still a functional change that I'd not make in a cleanup PR.

btw your change in switch_alignment_size is also a functional change. I don't think it would cause any problems and I think it's a good change. again, this is not something I'd do in a cleanup PR but I'd leave it up to you.

Reply:
> if it really makes you feel better

Sigh. It is not about "feeling better", it is about following good coding guidelines. There are two closely related functions, `same_large_alignment_p` and `switch_alignment_size`, which together ensure double alignment of objects. It makes no sense to use symbolic constant in one of them and numeric constant in another.

>  large alignment means 4-byte vs 8-byte

Feature `RESPECT_LARGE_ALIGNMENT` was introduced with the following comment: "Respect double alignment of objects during compaction", which means alignment to `2*DATA_ALIGNMENT` bytes.  Yes, we are currently using it only when `DATA_ALIGNMENT == 4`.  It does not mean it won't work with minimal changes for, say, `DATA_ALIGNMENT == 8`.

>  it's still a functional change that I'd not make in a cleanup PR

I don't quite understand this argument; in my opinion this sort of functional changes are OK when cleaning up.  I can rename the PR to "Clean up and small change..." if you think that is clearer.

Reply:
my concern about combining cleanup with functional changes isn't the commit naming, it is that it changes the amount of risk that the change introduces and thus the amount of testing that the change requires. for a cleanup only change we are asserting with code review alone that the code will produce the same output in all circumstances. for a functional change that is no longer the case. changing the implementation of switch_alignment_size to return a size that's smaller than min obj size could certainly cause problems that are not small (as I said, I don't actually think it would cause any problems but it's the kind of changes that have the potential to cause big problems, unlike the non functional changes). given the high bar for correctness in the GC you would need to do stress testing for this, and a small but not zero chance that the test will produce failures which then you need to be debug. IMO this is a considerable amount of additional work that you would undertake for a very limited benefit. my suggestion is that it would be more efficient to forgo the functional part of the change but it is your call.

Reply:
> changing the implementation of switch_alignment_size to return a size that's smaller than min obj size could certainly cause problems that are not small

Let's do this then and preserve release behavior:
```C++
size_t switch_alignment_size (BOOL already_padded_p)
{
#ifndef RESPECT_LARGE_ALIGNMENT
    assert (!"Should never be called");
#endif // RESPECT_LARGE_ALIGNMENT
   ...existing code...
}
```
I will scale down the changes and resubmit.